compiler/rustc_const_eval/src/transform/check_consts/check.rs

```diff
@@ -896,6 +900,20 @@ impl Visitor<'tcx> for Checker<'mir, 'tcx> {
                         }
                     }
 
+                    // const-eval of the `panic_display` fn assumes the argument is `&&str`
+                    if Some(callee) == tcx.lang_items().panic_display() {
+                        let is_str = match args[0].ty(&self.ccx.body.local_decls, tcx).kind() {
+                            ty::Ref(_, ty, _) => match ty.kind() {
+                                ty::Ref(_, ty, _) => ty.is_str(),
+                                _ => false,
+                            },
+                            _ => false,
+                        };
+                        if !is_str {
+                            self.check_op(ops::PanicNonStr);
+                        }
```

Comment:
Note that you could use `matches!()` here to turn this into a single `match` if you want:
```suggestion
                        match args[0].ty(&self.ccx.body.local_decls, tcx).kind() {
                            ty::Ref(_, ty, _) if matches!(ty.kind(), ty::Ref(_, ty, _) if ty.is_str()) => {}
                            _ => self.check_op(ops::PanicNonStr),
                        }
```